backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/ir/interop/IrProviderForInteropStubs.kt

```diff
@@ -0,0 +1,179 @@
+import org.jetbrains.kotlin.resolve.descriptorUtil.module
+
+/**
+ * Generates external IR declarations for descriptors from interop libraries.
+ * [isSpecialInteropCase] allows to delegate providing of interop symbols to other providers.
+ * For example, for CEnums we need to generate non-lazy IR.
+ */
+class IrProviderForInteropStubs(
+        private val isSpecialInteropCase: (IrSymbol) -> Boolean
+) : LazyIrProvider {
+
+    override lateinit var declarationStubGenerator: DeclarationStubGenerator
+
+    // We need to collect provided Objective-C classes to be able to generate RTTI for them.
+    // TODO: Unify with BuiltInFictitiousFunctionIrClassFactory?
+    private val filesMap = mutableMapOf<PackageFragmentDescriptor, IrFile>()
+
+    var module: IrModuleFragment? = null
+        set(value) {
+            if (value == null)
+                error("Provide a valid non-null module")
+            if (field != null)
+                error("Module has already been set")
+            field = value
+            value.files += filesMap.values
+        }
+
+    override fun getDeclaration(symbol: IrSymbol): IrLazyDeclarationBase? =
+            when {
+                // TODO: LazyIrProvider appears to be a bad interface.
+                //  Incoming symbol might be already bound and we need to return its current owner.
+                //  The problem is that it isn't necessary a LazyIr declaration.
+                //  So for now we relate on correct behavior of subsequent providers.
+                symbol.isBound -> null
+                isSpecialInteropCase(symbol) -> null
+                symbol.descriptor.module.isFromInteropLibrary() -> provideIrDeclaration(symbol)
+                else -> null
+            }
+
+    private fun provideIrDeclaration(symbol: IrSymbol): IrLazyDeclarationBase = when (symbol) {
+        is IrSimpleFunctionSymbol -> provideIrFunction(symbol)
+        is IrPropertySymbol -> provideIrProperty(symbol)
+        is IrTypeAliasSymbol -> provideIrTypeAlias(symbol)
+        is IrClassSymbol -> provideIrClass(symbol).also {
+            // We need to use extension for a descriptor because IR-based function involves
+            // LazyTable referencing.
+            if (it.descriptor.isObjCClass()) collectObjCClasses(it)
+        }
+        is IrConstructorSymbol -> provideIrConstructor(symbol)
+        is IrFieldSymbol -> provideIrField(symbol)
+        else -> error("Unsupported interop declaration: symbol=$symbol, descriptor=${symbol.descriptor}")
+    }
+
+    private fun referenceSuperClassifiers(descriptor: ClassDescriptor) {
+        val symbolTable = declarationStubGenerator.symbolTable
+        descriptor.getAllSuperClassifiers().forEach { symbolTable.referenceClassifier(it) }
+    }
+
+    private fun collectObjCClasses(klass: IrClass) {
+        val descriptor = klass.descriptor
+        // Force-load types to be able to generate RTTI for the given class.
```

Comment:
Is RTTI really required for Obj-C classes? It shouldn't be.

Reply:
IIUC, it is needed when we perform instance checking.